node/core/pvf/src/prepare/worker.rs

```diff
@@ -288,31 +281,43 @@ pub fn worker_entrypoint(socket_path: &str) {
 				worker_pid = %std::process::id(),
 				"worker: preparing artifact",
 			);
-			let artifact_bytes = prepare_artifact(&code).serialize();
 
-			// Write the serialized artifact into into a temp file.
-			tracing::debug!(
-				target: LOG_TARGET,
-				worker_pid = %std::process::id(),
-				"worker: writing artifact to {}",
-				dest.display(),
-			);
-			async_std::fs::write(&dest, &artifact_bytes).await?;
+			let bytes = match prepare_artifact(&code) {
+				Artifact::Error(err) => {
+					// Serialized error will be written into the socket.
+					err.encode()
+				},
+				Artifact::Compiled(compiled_artifact) => {
+					// Write the serialized artifact into a temp file.
+					// Since a compiled artifact can be heavy, we only send a single
+					// byte to indicate the success.
+					let artifact_bytes = compiled_artifact.encode();
+
+					tracing::debug!(
+						target: LOG_TARGET,
+						worker_pid = %std::process::id(),
+						"worker: writing artifact to {}",
+						dest.display(),
+					);
+					async_std::fs::write(&dest, &artifact_bytes).await?;
+
+					vec![1u8]
```

Comment:
I think sending `1u8` to indicate success doesn't work any longer.

To understand why, look at the `Artifact::Error(err)` case. `err` is of type `PrepareError` which has roughly the following shape:

```rust
#[derive(Debug, Clone, Encode, Decode)]
pub enum PrepareError {
	Prevalidation(String),
	Preparation(String),
	DidNotMakeIt,
}
```

`PrepareError` implements `Decode` via deriving and for enum that is very simple: each variant is assigned a special one byte tag value. Thus, encoding `PrepareError::Preparation` will start with a `1`.

I hope you can see the problem now.

Besides that, on the receiving side the code also looks a bit clunky. Can we just put a proper typed value on the wire, such as `Result<(), PrepareError>`? That would simultaneously solve this issue and also clean up the `match` on the receiving side.

Reply:
[agreed](https://github.com/paritytech/polkadot/pull/3907/commits/8bdc7d3791aec014a3f1dd26b2291add569bc07a)